proxy/nftables: reject packets destined for invalid ports of service ips

aroradaman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
NFTables proxier will reject packets destined for invalid ports(not specified) of service IPs.
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aroradaman commented:
Tried locally, traffic to invalid port get instantly rejected rather than blackholing.
```
<<K9s-Shell>> Pod: default/busybox-678fbdc4d9-gjfgf | Container: busybox 
root@busybox-678fbdc4d9-gjfgf:/# curl 10.96.104.46:80
Hostname: server-59d49d654f-wjrks
IP: 127.0.0.1
IP: ::1
IP: 10.244.2.2
IP: fe80::bc73:26ff:fe5f:5561
RemoteAddr: 10.244.2.3:37536
GET / HTTP/1.1
Host: 10.96.104.46
User-Agent: curl/7.88.1
Accept: */*

root@busybox-678fbdc4d9-gjfgf:/# curl 10.96.104.46:81 -v
*   Trying 10.96.104.46:81...
* connect to 10.96.104.46 port 81 failed: Connection refused
* Failed to connect to 10.96.104.46 port 81 after 0 ms: Couldn't connect to server
* Closing connection 0
curl: (7) Failed to connect to 10.96.104.46 port 81 after 0 ms: Couldn't connect to server
```

aroradaman commented:
/test pull-kubernetes-verify

danwinship commented:
> (traffic for VIPs has to be generated within the cluster)

The iptables and nftables modes both explicitly allow for the possible that the network implementation [might route traffic to clusterIPs from off-node](https://github.com/kubernetes/kubernetes/blob/v1.29.0/pkg/proxy/nftables/proxier.go#L1340-L1344)

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -601,6 +606,15 @@ func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+			Chain: kubeServicesChain,
+			Rule: knftables.Concat(
+				ipX, "daddr", "==", "{", strings.Join(proxier.serviceCIDRs.UnsortedList(), ","), "}",
+				"reject",
```
Sorry, I changed my mind about this while we were discussing it on slack, but I guess I wasn't clear :slightly_smiling_face: 

ISTM that traffic to an unused port on an active cluster IP should get rejected, in the same way that (assuming no firewall), traffic to a non-listening port on a host would get rejected.

But traffic to an unused cluster IP should just get dropped, in the same way that traffic to an unused IP on your local network would get dropped/ignored. (Right now we just ignore unassigned clusterIPs, which probably means they end up getting routed along the default route, which means they eventually get dropped by someone, but maybe not as efficiently as they should.)

So there should be two commits:
1. reject connections to inactive ports/protocols on active clusterIPs
2. drop packets to unused clusterIPs

danwinship commented on pkg/proxy/config/config.go:
```diff
@@ -371,3 +373,111 @@ func (c *NodeConfig) handleDeleteNode(obj interface{}) {
+	// OnServiceCIDRSynced is called once all the initial event handlers were
+	// called and the state is fully propagated to local cache.
+	OnServiceCIDRSynced()
+}
```
ServiceCIDRs aren't going to change much, and the full list of active ServiceCIDRs is always going to be short anyway. Just simplify this to `OnServiceCIDRsChanged()`, and have the `ServiceCIDRConfig` do the work of keeping the current list, and just pass a new list down every time it changes. (And then there's only 1 dummy method to add to the backends that aren't supporting this.)

(NodeConfig should have worked this way too...)

danwinship commented on pkg/proxy/config/config.go:
```diff
@@ -22,9 +22,11 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 	discovery "k8s.io/api/discovery/v1"
+	networking "k8s.io/api/networking/v1alpha1"
```
call this `networkingv1alpha1`; we might need to pull in `networkingv1` at some point too.

(yes, discovery is wrong)

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -601,6 +606,15 @@ func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+		tx.Add(&knftables.Rule{
+			Chain: kubeServicesChain,
+			Rule: knftables.Concat(
+				ipX, "daddr", "==", "{", strings.Join(proxier.serviceCIDRs.UnsortedList(), ","), "}",
```
"==" is optional in rules, and the "canonical" syntax is to omit it:

```
$ nft add chain ip test equals
$ nft add rule ip test equals 'ip daddr == { 1.2.3.4, 5.6.7.8 } drop'
$ nft list chain ip test equals
table ip test {
	chain equals {
		ip daddr { 1.2.3.4, 5.6.7.8 } drop
	}
}
```

danwinship replied:
This might be a good example of a rule that should have a comment explaining it...
(nftables comment, not a golang comment)

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -795,6 +809,42 @@ func (proxier *Proxier) OnNodeDelete(node *v1.Node) {
+	defer proxier.mu.Unlock()
+
+	for _, cidr := range serviceCIDR.Spec.CIDRs {
+		if proxier.ipFamily == v1.IPv6Protocol && isIPv6CIDRString(cidr) {
```
```suggestion
		if proxier.ipFamily == v1.IPv6Protocol && netutils.IsIPv6CIDRString(cidr) {
```

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -795,6 +809,42 @@ func (proxier *Proxier) OnNodeDelete(node *v1.Node) {
+
+// OnServiceCIDRUpdate is called whenever modification of an existing ServiceCIDR
+// object is observed.
+func (proxier *Proxier) OnServiceCIDRUpdate(_, _ *networking.ServiceCIDR) {}
```
you need to handle these methods too

aroradaman replied:
yes, was prioritizing to get core logic (nft part) right first.

danwinship commented:
> If this doesn't work for older kernels < 5.9, maybe we can add this rule at the end of the `nat-output` chain

`nat-output` is only for traffic that originates from a host-network process. (See https://github.com/kubernetes/kubernetes/pull/122687/files, which finally tries to document all this clearly.)

Ignoring refactoring for a moment, to handle just the unused ports/IPs:
- We don't need to do anything on "input" because that's only for traffic which is terminating at a local IP, and this traffic is both (a) not terminating, and (b) not addressed to a local IP.
- We need to have a rule to drop/reject on "forward", and it doesn't matter what the priority is, because any "good" service traffic would already have been rewritten in "prerouting", so any clusterIP-destined traffic that reaches "forward" is definitely bad.
- We need to have a rule to drop/reject on "output", to handle the case of "hostNetwork process connecting to bad service IP", and in that case the priority of the chain _does_ matter, because we have to make sure the drop/reject rules run after the DNAT rules that are also on "output". So this means you can't just reuse the same "filter-output" base chain that LoadBalancerSourceRanges handling uses, because those rules have to run _before_ DNAT... So we'll need a new base chain on filter and output, but with a higher-than-DNAT priority rather than a lower-than-DNAT priority.

#122687 also suggests we might want to refactor the "no endpoints" drop/reject rules as part of this, which might make sense? In that case, note that the full set of what we need to do where is:

- "input": Drop/reject connections to `@no-endpoint-nodeports`
- "forward": Drop/reject connections to `@no-endpoint-services`. Reject connections to unused ports on active clusterIPs. Drop connections to unused clusterIPs.
- "output": Drop/reject connections to `@no-endpoint-services`. Reject connections to unused ports on active clusterIPs. Drop connections to unused clusterIPs.
- "postrouting": ALTERNATE: we could do the "Drop connections to unused clusterIPs" part here rather than doing it in both "forward" and "output". (But we can't call `reject` from here just like we can't call it from `prerouting`.)

aroradaman commented on pkg/proxy/nftables/proxier_test.go:
```diff
@@ -1227,18 +1247,34 @@ func TestNodePorts(t *testing.T) {
 }
 
 func TestDropInvalidRule(t *testing.T) {
```
This test will be dropped eventually, added just for CI.

aroradaman commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml:
```diff
@@ -1183,6 +1183,13 @@ items:
     verbs:
     - list
     - watch
+  - apiGroups:
```
not sure how to fix/add this in kind/ci

aroradaman commented:
/test pull-kubernetes-e2e-kind-nftables 
/test pull-kubernetes-unit

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -336,6 +340,7 @@ var nftablesBaseChains = []nftablesBaseChain{
 	{"filter-input", knftables.FilterType, knftables.InputHook, knftables.DNATPriority + "-10"},
 	{"filter-forward", knftables.FilterType, knftables.ForwardHook, knftables.DNATPriority + "-10"},
 	{"filter-output", knftables.FilterType, knftables.OutputHook, knftables.DNATPriority + "-10"},
+	{"filter-output-post-dnat", knftables.FilterType, knftables.OutputHook, knftables.FilterPriority},
```
(Hm... I don't like the fact that it's not really obvious that "FilterPriority" is after "DNATPriority"... maybe I just switch to all numeric priorities...)

aroradaman replied:
I think it is fine, anyone working on NFT should be aware of conntrack, dnat, filter, snat priorities ...

(random thought: maybe we should do filtering even before conntrack, to apply it on runtime  (already established connections) 😅 )

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -336,6 +340,7 @@ var nftablesBaseChains = []nftablesBaseChain{
 	{"filter-input", knftables.FilterType, knftables.InputHook, knftables.DNATPriority + "-10"},
 	{"filter-forward", knftables.FilterType, knftables.ForwardHook, knftables.DNATPriority + "-10"},
 	{"filter-output", knftables.FilterType, knftables.OutputHook, knftables.DNATPriority + "-10"},
+	{"filter-output-post-dnat", knftables.FilterType, knftables.OutputHook, knftables.FilterPriority},
```
We may want to reconsider the chain names. At any rate, at this point you need to remove the comment on line 328 that claims that `All of our base chains have names that are just "${type}-${hook}".`

aroradaman replied:
for consistency should we follow ${type}-${hook}-${priority} ?

danwinship replied:
that makes sense, except for the fact that most of the priorities are negative numbers so you'd end up with double hyphens... but something along those lines might be good. You could just use "early" "normal" "late" or something like that?

aroradaman replied:
how about "${type}-${hook}-(before|at|after)-${target}" ? eg. `filter-postrouting-after-dnat` and `filter-prerouting-before-dnat`

"early" "normal" "late" will revolve around `dnat`, we also use `snat` priority for nat-postrouting.
Maybe we can define priority around a target - snat | dnat ?

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -88,6 +91,7 @@ const (
 
 	// chains for special filtering rules
 	kubeForwardChain     = "forward"
+	kubeOutputChain      = "output"
```
same needs a better name or else doesn't need to exist comment

danwinship commented on pkg/proxy/nftables/proxier_test.go:
```diff
@@ -574,6 +582,7 @@ func TestOverallNFTablesRules(t *testing.T) {
 		add rule ip kube-proxy endpoint-SGOXE6O3-ns2/svc2/tcp/p80__10.180.0.2/80 ip saddr 10.180.0.2 jump mark-for-masquerade
 		add rule ip kube-proxy endpoint-SGOXE6O3-ns2/svc2/tcp/p80__10.180.0.2/80 meta l4proto tcp dnat to 10.180.0.2:80
 
+		add element ip kube-proxy cluster-ips { 172.30.0.42 }
```
be consistent about whether you add this before or after the service-ips lines

danwinship commented on pkg/proxy/nftables/proxier_test.go:
```diff
@@ -879,14 +892,14 @@ func TestClusterIPGeneral(t *testing.T) {
 			destIP:   "172.30.0.42",
 			destPort: 80,
-			output:   "",
+			output:   "REJECT",
```
YAY but this reminds me there should be packet flow tests for the drop rule too (I think the stuff in https://github.com/kubernetes/kubernetes/commit/4552b2d84c8f0a7c87d29ff683ccb19470468b2b will handle processing that rule; what eventually got merged only handles the "!= { ... }" case, but if you add back the non-!= case it should handle your rule)

aroradaman commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -795,6 +815,33 @@ func (proxier *Proxier) OnNodeDelete(node *v1.Node) {
+		if proxier.ipFamily == v1.IPv6Protocol && !isIPv4CIDR {
+			cidrsForProxier = append(cidrsForProxier, ipNet.String())
+		}
+	}
```
Can I also assume that cidrs will be canonical?
trying to avoid https://github.com/kubernetes/kubernetes/issues/122611

danwinship replied:
> Can I also assume that cidrs will be canonical?

yes: https://github.com/kubernetes/kubernetes/blob/v1.29.0/pkg/apis/networking/validation/validation.go#L766

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -599,6 +599,11 @@ func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+	tx.Add(&knftables.Rule{
+		Chain: kubeEndpointsCheckChain,
+		Rule:  knftables.Concat("jump", kubeServiceEndpointsCheckChain),
+	})
```
might be cleaner to have `service-endpoints-check` and `nodeport-endpoints-check` and have both linked to `filter-input` but only the former linked to `filter-forward` and `filter-output`?

aroradaman commented on pkg/proxy/nftables/helpers_test.go:
```diff
@@ -591,10 +593,18 @@ func tracePacket(t *testing.T, nft *knftables.Fake, sourceIP, protocol, destIP,
 	// the destination IP that later rules match against).
 	if len(tracer.outputs) != 0 {
-		destIP = strings.Split(tracer.outputs[0], ":")[0]
+		addr, err := net.ResolveTCPAddr("tcp", tracer.outputs[0])
```
I was facing some issues with parsing IP:PORT after adding the tracing on`filter-input` chain. It failed to parse IPv6 addr with `:` in between ... 

Will cross-check again.

danwinship replied:
Oh, right. Yeah, I think the iptables tracer (which this code is copied from) was only ever used for IPv4 tests. You want `net.SplitHostPort` though, not `net.ResolveTCPAddr`.

aroradaman replied:
updated, this came after removing destIP local check (`fib daddr type local`) from `node-port-endpoints-check` chain

danwinship commented:
Oh, also, you need to update the "kube-proxy's use of nftables hooks" section of the new `pkg/proxy/nftables/README.md` to reflect your changes here.

tnqn commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -429,6 +450,18 @@ func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+	// Drop traffic to unused ClusterIPs.
+	if len(proxier.serviceCIDRs) > 0 {
+		tx.Add(&knftables.Rule{
+			Chain: KubeFilterPostRouting,
+			Rule: knftables.Concat(
+				ipX, "daddr", "{", proxier.serviceCIDRs, "}",
+				"drop",
+			),
+			Comment: ptr.To("Drop traffic to unused ClusterIPs"),
```
1. Other kinds of drop/reject rules are installed in "regular chains", as opposed to "base chains", I wonder if this should follow the convention.
2. Given that we have to install reject rules in forward and output chains, is it better to install rules of "reject invalid ports of ClusterIPs" and "drop unused ClusterIPs" in the regular chain, to which "filter-forward" and "filter-output-post-dnat" will jump? Something like the following:
```
add rule ip kube-proxy filter-forward ct state new jump service-endpoints-check
add rule ip kube-proxy filter-forward ct state new jump clusterip-check

add rule ip kube-proxy filter-output-post-dnat ct state new jump clusterip-check

add rule ip kube-proxy clusterip-check ip daddr @cluster-ips reject
add rule ip kube-proxy clusterip-check ip daddr { 10.96.0.0/16 } drop
```
Also, it may be a bit more efficient to drop traffic to unused ClusterIPs earlier in forward or output hook.

cc @danwinship for his opinion.

aroradaman replied:
> Other kinds of drop/reject rules are installed in "regular chains", as opposed to "base chains", I wonder if this should follow the convention.

https://github.com/kubernetes/kubernetes/pull/122692#discussion_r1451580776

danwinship replied:
I had suggested this approach above: https://github.com/kubernetes/kubernetes/pull/122692#discussion_r1451580776

aroradaman replied:
I like this idea of keeping cluster-ip checks together in one chain

aroradaman commented on cmd/kube-proxy/app/server.go:
```diff
@@ -933,6 +933,16 @@ func (s *ProxyServer) Run() error {
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+		serviceCIDRConfig, err := config.NewServiceCIDRConfig(informerFactory.Networking().V1alpha1().ServiceCIDRs(), s.Config.ConfigSyncPeriod.Duration)
+		if err != nil {
+			s.logger.Error(err, "failed to create serviceCIDR config for watches")
+			return err
+		}
```
Yes, error is almost irrelevant here.
I had to add it to pass `pull-kubernetes-linter-hints` job which was failing.

tnqn commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -531,7 +531,7 @@ func ClusterRoles() []rbacv1.ClusterRole {
 		rbacv1helpers.NewRule("list", "watch").Groups(legacyGroup).Resources("services", "endpoints").RuleOrDie(),
 		rbacv1helpers.NewRule("get", "list", "watch").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
-
+		rbacv1helpers.NewRule("list", "watch").Groups(networkingGroup).Resources("servicecidrs").RuleOrDie(),
```
Reminded by liggitt's comment in https://github.com/kubernetes/kubernetes/pull/122837#issuecomment-1898680505, adding this permission should be feature gated. service-cider-controller feature gate it: https://github.com/kubernetes/kubernetes/blob/8835dabc3bb38aaedaccf6a0237c1dabac17f0d8/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go#L372-L383

aroradaman replied:
will feature gate this then.

tnqn reviewed (commented):
Code LGTM, should there be an unit test validating specific traffic gets rejected/dropped?

tnqn reviewed (commented):
/lgtm
The commit message of the first commit is not wrapped, better to adjust it following  https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#wrap-the-commit-message-body-at-72-characters

danwinship commented on pkg/proxy/config/config.go:
```diff
@@ -371,3 +375,97 @@ func (c *NodeConfig) handleDeleteNode(obj interface{}) {
+		cidrs:        sets.New[string](),
+	}
+
+	_, err := serviceCIDRInformer.Informer().AddEventHandlerWithResyncPeriod(
```
The linter doesn't like the way the other informer setup code is doing it because it looks like maybe the author of the code didn't even realize the function could return an error. It ought to pass if you _explicitly_ ignore the error:

    _, _ = serviceCIDRInformer.Informer().AddEventHandlerWithResyncPeriod(

Maybe fix the other informer setups to do the same thing, for consistency, and so someone else doesn't get confused by linter errors when refactoring that code later. (I'd do that as a separate commit before the main work.)

danwinship commented on pkg/proxy/nftables/README.md:
```diff
@@ -64,6 +64,10 @@ Kube-proxy uses nftables for four things:
+
+  - Rejecting packets to undefined ports of ClusterIPs.
 
 This is implemented as follows:
```
need to update this part too before this merges

danwinship commented on pkg/proxy/nftables/helpers_test.go:
```diff
@@ -202,11 +202,12 @@ func (tracer *nftablesTracer) addressMatches(ipStr, not, ruleAddress string) boo
 	for _, address := range addresses {
-		if tracer.addressMatches(ipStr, "", address) {
+		match := tracer.addressMatches(ipStr, "", address)
+		if (match && not == "!= ") || (!match && not == "") {
```
oof. maybe change the `not string` arg to `wantMatch bool` (which the caller would just pass `not == ""` for) and then you can say `if match == wantMatch {`

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -63,6 +64,16 @@ const (
+	kubeFilterInput          = "filter-input"
+	kubeFilterForward        = "filter-forward"
+	kubeFilterOutput         = "filter-output"
+	KubeFilterOutputPostDNAT = "filter-output-post-dnat"
```
you have a capital "K" there

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -71,12 +82,19 @@ const (
 
+	// handling traffic to unused ClusterIPs and invalid
+	// ports of ClusterIPs
+	KubeClusterIPsCheckChain = "cluster-ips-check"
```
likewise

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -63,6 +64,16 @@ const (
 	kubeProxyTable = "kube-proxy"
 
+	// base chains
+	kubeFilterPreRouting     = "filter-prerouting"
```
"Prerouting" (and "Postrouting"); be consistent between the two forms. If it was `filter-pre-routing` then `FilterPreRouting` would be right, but it's not.

danwinship commented on pkg/proxy/nftables/proxier.go:
```diff
@@ -63,6 +64,16 @@ const (
+	KubeFilterOutputPostDNAT = "filter-output-post-dnat"
+	kubeNATPreRouting        = "nat-prerouting"
+	kubeNATOutput            = "nat-output"
+	kubeNATPostRouting       = "nat-postrouting"
```
For consistency with the other chains, these constants should all have `Chain` at the end of their name (eg `kubeFilterInputChain`)

(Also, I'm realizing that the reason the existing constants have names starting with `kube` is because in the iptables proxy, the actual chain names all started with `KUBE` too... I probably should have renamed them in nftables... You do not need to fix this, but you can if you want. If you do, you should fix the existing constants first as a separate commit.)

danwinship commented on pkg/proxy/nftables/proxier_test.go:
```diff
@@ -494,6 +494,9 @@ func TestOverallNFTablesRules(t *testing.T) {
 	)
 
+	// populate serviceCIDRs
+	fp.OnServiceCIDRsChanged([]string{"172.30.0.0/24", "fd00:10:96::/112"})
```
ok, super nit-picky, but a `/16` IPv4 service cidr would be a lot more normal

also, maybe just set `proxier.serviceCIDRs` explicitly in `NewFakeProxier` rather than doing this in each unit test?

aroradaman committed:
pkg/proxy/config: use blank identifier instead of ignoring error

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/config: rename import aliases

* coreinformers -> v1informers
* discovery -> discoveryv1
* discoveryinformers -> discoveryv1informers

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/nftables: refactor packet tracer address matching

Use bool instead of not-equal-operator as string in
tracer.addressMatches of helpers_test.go

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/nftables: rename constant names for nftable objects

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/nftables: handle traffic to cluster ip

NFTables proxy will now drop traffic directed towards unallocated
ClusterIPs and reject traffic directed towards invalid ports of
Cluster IPs.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/nftables: handle traffic to node ports with no endpoints

NFTables proxy will no longer install drop and reject rules for node
port services with no endpoints in chains associated with forward and
output hooks.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
/test pull-kubernetes-e2e-kind

danwinship commented:
so much refactoring!
/lgtm
/approve

liggitt commented:
/approve

for conditional list / watch permission on the network resource

aroradaman commented:
/kind cleanup

aroradaman commented:
@danwinship should we add a release note for cluster IP drop/reject?

aroradaman commented:
/release-note-none
